dynatrace_topology/stackstate_checks/dynatrace_topology/dynatrace_topology.py

```diff
@@ -0,0 +1,276 @@
+            raise ConfigurationError('Missing URL in configuration.')
+        if 'token' not in instance:
+            raise ConfigurationError('Missing API Token in configuration.')
+
+        self.url = instance.get('url')
+        self.token = instance.get('token')
+        self.tags = instance.get('tags', [])
+        self.domain = instance.get('domain', '')
+        self.environment = instance.get('environment', 'production')
+        self.verify = instance.get('verify', True)
+        self.cert = instance.get('cert', '')
+        self.keyfile = instance.get('keyfile', '')
+
+        try:
+            self.start_snapshot()
+            self.process_topology()
+            self.stop_snapshot()
+            msg = "Dynatrace topology processed successfully"
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=self.tags, message=msg)
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.tags, message=str(e))
+
+    def process_topology(self):
+        """
+        Collects components and relations for each component type from dynatrace smartscape topology API
+        """
+        start_time = datetime.now()
+        self.log.info("Starting the collection of topology")
+        services = self.collect_services()
+        processes = self.collect_processes()
+        processes_groups = self.collect_process_groups()
+        hosts = self.collect_hosts()
+        applications = self.collect_applications()
+        topology = {"service": services, "process": processes, "process-group": processes_groups, "host": hosts,
+                    "application": applications}
+        # process component for each type
+        for comp_type, response in topology.items():
+            self.process_component(response, comp_type)
+        end_time = datetime.now()
+        time_taken = end_time - start_time
+        self.log.info("Time taken to collect the topology is: {} seconds".format(time_taken.total_seconds()))
+
+    def collect_relations(self, component, externalId):
+        """
+        Collects relationships from different component-types
+        :param component: the component for which relationships need to be extracted and processed
+        :param externalId: the component externalId for and from relationship will be created
+        :return: None
+        """
+        outgoing_relations = component.get("fromRelationships", {})
+        incoming_relations = component.get("toRelationships", {})
+        for relation_type in outgoing_relations.keys():
+            # Ignore `isSiteOf` relation since location components are not processed right now
+            if relation_type != "isSiteOf":
+                for target_id in outgoing_relations[relation_type]:
+                    self.relation(externalId, target_id, relation_type, {})
+        for relation_type in incoming_relations.keys():
```

Comment:
If I believe Zandre, the agent is not going to help you dedup this. I also do not believe we can trust the batcher to not send a batch within the same check run, so the only thing we can do is make sure we have a unique `externalId` for the relation. This should ensure that StackState dedups the relations.